notes/conds.md

```diff
@@ -0,0 +1,84 @@
+| ------------ | --------------------------------- | ------------------------------------------ | ---------------------------------------------- |
+| `/${string}` |`[] -> Any`                        | Selects value using JSON Pointer            | `["/foo/0"]`                                  |
+|              |`[] -> Any`                        | Selects whole value like "" in JSON Pointer | `[]`                                          |
+| `==`         |`[Exp<Any>, Exp<Any>] -> Bool`     | Equality                                    | `["==", ["/size"], 1]`                        |
+| `<`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than                                   | `["<",  ["/size"], 1024]`                     |
+| `<=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Less than or equal                          | `["<=", ["/size"], 1024]`                     |
+| `>`          |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than                                | `[">", ["@", "/size"], 0]`                    |
+| `>=`         |`[Exp<Num>, Exp<Num>] -> Bool`     | Greater than or equal                       | `[">=", ["@", "/size"], 32]`                  |
+| `not`        |`[Expr<Bool>] -> Bool`             | Negation                                    | `["not", ["<=", ["/size"], 128]`              |
+| `some`       |`[Expr<Any[]>, Expr<Bool>] -> Bool`| Collection predicate                        | `["some", ["/to"], ["==", [], "hi@web.mail"]]`|
+| `match`      |`[Expr<Str>, Exp<Str>] -> Bool`    | String wildcard                             | `["match", ["/to"], "*@foo.com"]`             |
+| `or`         |`Exp<Bool>[] -> Bool`              | Logical **or** combinator                   | `["or", [">", ["/n"], 0], ["==", ["/n"], 0]]` |
+| `and`        |`Exp<Bool>[] -> Bool`              | Logical **and** combinator                  | `["and", [">", ["/n"], 0], ["<", ["/n"], 10]]`|
+
+## Examples
+
+Delegates `email/send` command on `did:key:alice` with a following constraint: Email is not send to address on `gmail.com`, or someone from `fission.codes` should be cc-ed.
+
+```js
+{
+  "sub": "did:key:alice",
+  "iss": "did:key:alice",
+  "aud": "did:key:bob",
+  "cmd": "email/send",
+  "cond": [
+    ["or",
+      // no recipient should be at `@gmail.com`
+      ["every", ["/to"], ["not", ["match", [], "*@gmail.com"]]],
+      // or cc should include @fission.codes
+      ["some", ["/cc"], ["match", [], "*@fission.codes"]]
+    ]
+  ]
+}
+```
+
+## logic programming language
+
+We are also considering logic programming alternative inspired by datomic query styntax.
+
+### concrete proposal for the inital release
+
+Here's a sketch that shows the features
+
+```js
+{
+  "iss": "did:key:alice",
+  // ...
+  "policy": [
+    ["var", ["?x", "?y"], // The `?` here is just by convention
+      ["args", "foo..bar.[].baz", "?x"],
```

Comment:
```suggestion
      ["args", "foo..bar.[].baz", ["?x"]],
```

I would prefer something like this so it's distinct from the string literal. alternatively it could use

```suggestion
      ["args", "foo..bar.[].baz", {"?": "x"}],
```

I can also be convinced that we could keep as is and require escaping in strings

Reply:
This still looks like an object literal, though: `{"?": "x"}` (which I think is a totally fair thing to match on)

The idea is that since users get to define the variable, they can avoid namespace conflicts. It does mean that when looking at a string in this position, it's `variable | lit_string`, and you have to consult your table.

By introducing these variables in their own clause at the top, we avoid the confusion of "are we introducing a variable or doing pattern matching?"

Reply:
> The idea is that since users get to define the variable, they can avoid namespace conflicts. It does mean that when looking at a string in this position, it's `variable | lit_string`, and you have to consult your variable table.

Do you mean that `"?x"` in `["args", "foo..bar.[].baz", "?x"]` would be string literal if you have not declared `x` var and be a reference otherwise ? If so I would find it very confusing. I would much rather say that strings starting with `?` are variable references and if you do want to have string literal starting with `?` you need to have to escape it.

> This still looks like an object literal, though: {"?": "x"} (which I think is a totally fair thing to match on)

I was assuming that values (third element of triple) are always either variable or a constant (primitive type), so it would not collide with array or object hence my suggestion.

Reply:
> and if you do want to have string literal starting with ? you need to have to escape it.

Sure, that works 👍